modules/core/src/smithy4s/internals/DocumentEncoderSchemaVisitor.scala

```diff
@@ -0,0 +1,296 @@
+import smithy4s.schema.Primitive.PUUID
+import smithy4s.schema.Primitive.PDouble
+import smithy4s.schema.Primitive.PLong
+import smithy4s.schema.Primitive.PString
+
+trait DocumentEncoder[A] { self =>
+
+  def canBeKey: Boolean
+
+  def apply: A => Document
+  def contramap[B](f: B => A): DocumentEncoder[B] =
+    new DocumentEncoder[B] {
+
+      def apply: B => Document = f andThen self.apply
+
+      def canBeKey: Boolean = self.canBeKey
+    }
+
+}
+
+object DocumentEncoder {
+  // to remove
+  type DocumentEncoderMake[A] = Hinted[DocumentEncoder, A]
+
+  implicit val contraInstance: Contravariant[DocumentEncoder] =
+    new Contravariant[DocumentEncoder] {
+      def contramap[A, B](fa: DocumentEncoder[A])(
+          f: B => A
+      ): DocumentEncoder[B] = fa.contramap(f)
+    }
+}
+
+object DocumentEncoderSchemaVisitor extends SchemaVisitor[DocumentEncoder] {
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): DocumentEncoder[P] = tag match {
+    case PShort      => from(short => DNumber(BigDecimal(short.toInt)))
+    case PBigInt     => from(bigInt => DNumber(BigDecimal(bigInt)))
+    case PBoolean    => from(DBoolean(_))
+    case PByte       => from(byte => DNumber(BigDecimal(byte.toInt)))
+    case PBigDecimal => from(DNumber(_))
+    case PInt        => from(int => DNumber(BigDecimal(int)))
+    case PBlob =>
+      from(bytes => DString(Base64.getEncoder().encodeToString(bytes.array)))
+    case PUnit => fromNotKey(_ => DObject(Map.empty))
+    case PTimestamp =>
+      new DocumentEncoder[Timestamp] {
+        def canBeKey: Boolean = true
+
+        def apply: Timestamp => Document =
+          hints
+            .get(TimestampFormat)
+            .getOrElse(TimestampFormat.DATE_TIME) match {
+            case DATE_TIME     => ts => DString(ts.format(DATE_TIME))
+            case HTTP_DATE     => ts => DString(ts.format(HTTP_DATE))
+            case EPOCH_SECONDS => ts => DNumber(BigDecimal(ts.epochSecond))
+          }
+      }
+    case PDocument => fromNotKey(identity)
+    case PFloat    => from(float => DNumber(BigDecimal(float.toDouble)))
+    case PUUID     => from(uuid => DString(uuid.toString()))
+    case PDouble   => from(double => DNumber(BigDecimal(double)))
+    case PLong     => from(long => DNumber(BigDecimal(long)))
+    case PString   => from(DString(_))
+  }
+
+  override def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: CollectionTag[C],
+      member: Schema[A]
+  ): DocumentEncoder[C[A]] = {
+    val encoderS = apply(member)
+    fromNotKey[C[A]](c =>
+      DArray(tag.iterator(c).map(encoderS.apply).toIndexedSeq)
+    )
+  }
+
+  override def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): DocumentEncoder[Map[K, V]] = {
+    val keyEncoder = apply(key)
+    val valueEncoder = apply(value)
+    fromNotKey[Map[K, V]] { map =>
```

Comment:
Do you mind if I do these updates in a subsequent PR? There are a few things that could be re-written. I saw some of them but I told myself: let's do the mechanical translation first (remove all instances of schematic) and then we'll improve the visitors.

it will reduce the review overhead. you can consider this PR as a one-to-one translation (mostly, because some signature are different and so some code has to change)